### DISCUSSION TITLE
FrankenPHP

### DISCUSSION THREAD

**Discussion by SimJoSt** [deployphp/deployer#3902] [Features]
FrankenPHP (https://frankenphp.dev/) seems to come up more and more in all kinds of PHP hosting environments. I was wondering if anyone has played around with it already and as Deployer is using Caddy anyway, the jump to use it might not be that far.

**Comment by alexander-schranz** [comment]
I also think it would make sense to use frankenphp as caddy is already used. Only thing which may is that FrankenPHP requires newer PHP versions (8.2, 8.3, 8.4).

**Reply by antonmedv** [reply]
Instead of php-fpm?

**Reply by alexander-schranz** [reply]
@antonmedv Yes, Frankenphp is all at one PHP + Webserver (Caddy), it is offical [supported PHP SAPI](https://github.com/php/php-src/blob/4122daa49421b0d079857d669b041e0f23f328ff/ext/opcache/ZendAccelerator.c#L2852) and modern alternative to Apache+PHPMod and Nginx+PHP-FPM, it uses caddy webserver as a base and compiles PHP directly into its binary.

It also has support for a called "worker mode" which is very similar how Roadrunner and Swoole works but that is optional and configured and enabled in the caddy file.

**Reply by alexander-schranz** [reply]
FrankenPHP will be now part of the PHP foundation: https://externals.io/message/127347

**Reply by SimJoSt** [reply]
@antonmedv is the limitation of only supporting PHP 8.2+ an issue for use in deployer?
In one year PHP 8.2 will not receive any security updates any more, and it might be a good time to drop support for it, allowing for a switch.
What is your view on that?

**Reply by antonmedv** [reply]
Yes, we can deprecate php8.2 support.  Can we pull usage statistics somethere?

**Reply by alexander-schranz** [reply]
8.2 I think is supported by FrankenPHP and deployer 8 already requires 8.2 or?

Stats for the projects using the composer package and not the phar can be found here: 

For all versions: https://packagist.org/packages/deployer/deployer/php-stats
For Deployer 7: https://packagist.org/packages/deployer/deployer/php-stats#7 which may be the most interesting version currently.

**Reply by SimJoSt** [reply]
FrankenPHP always contains (only) one specific PHP version, as far as I understand. The current Debian version (and other prepackaged binaries) contains PHP 8.4.
If you have multiple applications on the same server the suggestion is to run all projects compatible with the version contained in FrankenPHP directly. All other projects with other PHP version requirements through fastcgi, like it's done in normal caddy.
Source: https://github.com/php/frankenphp/discussions/721#discussioncomment-15499775

Because of the challenges of this coupling between webserver and contained PHP version, the FrankenPHP team already deprecated the current install through debian package with warnings, etc. and is moving to another way of installing.
The docs to reflect that are currently in review and should be available soon.
In the future it will be easy to define the PHP version that should be contained in FrankenPHP, while the install and updates still works through `apt`. Probably, a script to manage package repositories.

How the "default" PHP version that will be part of FrankenPHP should be defined in deployer, I don't know.
Maybe it's just defined by the deployer version and only changed with major versions jumps.
User-defined might be too complicated.

**Reply by alexander-schranz** [reply]
Would not make deployer version restricted to a specific PHP Version. We use kind of mono repositories where deployer is used to deploy to different servers. The servers are PHP 8.2-8.4 the project is developed in a way it supports that versions.

I think as provision is mostly done once (?), deployer could really ask for the version the project need to run in if it can not detect it maybe for by check a composer.lock or composer.json file.

**Reply by SimJoSt** [reply]
As FrankenPHP is a full Caddy web server, it would be possible to just replace the binary and nothing would change.
Everybody who want to opt-in to the FrankenPHP features, with the specific PHP version that binary supports, can change the relevant Caddyfile for that specific project. Depending on the feature set people want to use from FrankenPHP, more config changes are necessary anyway.
So I think it could be a drop-in replacement, without any changes forced onto the user, because fastcgi is still used.
But it allows anybody interested, in going one step further.

If there is a smart way for deployer to orchestrate and provision more of FrankenPHP-specific features, that could also come later done the line.
